### PR TITLE
jcache support for spring cache manager

### PR DESCRIPTION
**Pull request by mustafaiman**
Spring applications can use Hazelcast's JCache support for caching. With this PR one can create a `javax.cache.CacheManager` bean backed by Hazelcast in their Spring xml configuration and link it with `SpringJCacheCacheManager`. This way Spring internally uses Hazelcast's JCache implementations for caching.

Related issue: https://github.com/hazelcast/hazelcast/issues/6036

Sample application-context.xml:
```xml
    <cache:annotation-driven cache-manager="cacheManager" />

    <hz:hazelcast id="instance">
        <hz:config>
            <hz:group name="dev" password="dev-pass"/>
            <hz:network port="5701" port-auto-increment="false">
                <hz:join>
                    <hz:multicast enabled="true"/>
                    <hz:tcp-ip enabled="false">
                        <hz:interface>127.0.0.1:5701</hz:interface>
                    </hz:tcp-ip>
                </hz:join>
                <hz:interfaces enabled="true">
                    <hz:interface>127.0.0.1</hz:interface>
                </hz:interfaces>
            </hz:network>
            <hz:cache name="name" />
            <hz:cache name="city" />
        </hz:config>
    </hz:hazelcast>

    <hz:cache-manager id="hazelcastJCacheCacheManager" instance-ref="instance" name="hazelcastJCacheCacheManager"/>

    <bean id="cacheManager" class="org.springframework.cache.jcache.JCacheCacheManager">
        <constructor-arg ref="hazelcastJCacheCacheManager" />
    </bean>

    <bean id="dummy" class="com.hazelcast.spring.cache.TestJCacheCacheManager.DummyBean" />
</beans>
```

### CODE REVIEW

**Comment by kobalski** [comment]
:+1:

File: hazelcast-spring/pom.xml [hunk 1]
```diff
@@ -30,7 +30,7 @@
     </parent>
 
     <properties>
-        <org.springframework.version>3.1.0.RELEASE</org.springframework.version>
+        <org.springframework.version>4.0.2.RELEASE</org.springframework.version>
```
Comment (jerrinot) [inline hazelcast-spring/pom.xml]: why is this needed? does it mean we no longer support Spring Framework 3.x ?

File: hazelcast-spring/src/main/java/com/hazelcast/spring/CacheManagerBeanDefinitionParser.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+        public BeanDefinitionBuilder getBuilder() {
+            return this.builder;
+        }
+
+        public void handle(Element element) {
+            handleCommonBeanAttributes(element, builder, parserContext);
+            final NamedNodeMap attrs = element.getAttributes();
+            if (attrs != null) {
+                Node instanceRefNode = attrs.getNamedItem("instance-ref");
```
Comment (jerrinot) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/CacheManagerBeanDefinitionParser.java]: is this the only way to pass the instance? Is there any other way to pass the instance? I do not know spring that well, I'm just asking.

File: hazelcast-spring/pom.xml (hunk 1 above)
Comment (mustafaiman) [inline hazelcast-spring/pom.xml]: JCacheCacheManager is present in 3.2.0 at least. However Spring 3.2.0 depends on JCache 0.5.
Spring 4.0.2 both have JCacheCacheManager and depends on JCache 1.0.0. This extension works under Spring 4 and JCache 1.0. It may work with Spring 3.x and JCache 0.5 but I could not test it yet.
Rest of the Spring module should work fine with Spring 3.x

File: hazelcast-spring/src/main/java/com/hazelcast/spring/CacheManagerBeanDefinitionParser.java (hunk 2 above)
Comment (mustafaiman) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/CacheManagerBeanDefinitionParser.java]: This is how all the other custom beans of Hazelcast (`<hz:map> , <hz:list>, etc`) passes the instance reference. I do not know any other way to do that.

File: hazelcast-spring/pom.xml (hunk 1 above)
Comment (mustafaiman) [inline hazelcast-spring/pom.xml]: I have just tested it. It works with Spring 3.2.0 and JCache 0.5. We should mention in documentation that those who want to use this feature should depend on JCache 0.5.

Comment (jerrinot) [inline hazelcast-spring/pom.xml]: can we make a runtime check and show a hint to a user?

File: hazelcast-spring/src/main/java/com/hazelcast/spring/CacheManagerBeanDefinitionParser.java (hunk 2 above)
Comment (jerrinot) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/CacheManagerBeanDefinitionParser.java]: ok, as long as its consistent with the rest I'm fine with it.

File: hazelcast-spring/src/test/java/com/hazelcast/spring/cache/TestJCacheCacheManager.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+import javax.annotation.Resource;
+import javax.cache.CacheManager;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+@RunWith(CustomSpringJUnit4ClassRunner.class)
+@ContextConfiguration(locations = {"jCacheCacheManager-applicationContext-hazelcast.xml"})
+@Category(QuickTest.class)
+public class TestJCacheCacheManager {
```
Comment (jerrinot) [inline hazelcast-spring/src/test/java/com/hazelcast/spring/cache/TestJCacheCacheManager.java]: that's an unusual naming. why not `JCacheCacheManagerTest` ?

File: hazelcast-spring/src/test/java/com/hazelcast/spring/cache/TestJCacheCacheManager.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    @Resource(name = "cacheManager")
+    private JCacheCacheManager springCacheManager;
+
+    @Resource(name = "cacheManager2")
+    private CacheManager cacheManager2;
+
+    @BeforeClass
+    @AfterClass
+    public static void start() {
+        Hazelcast.shutdownAll();
```
Comment (jerrinot) [inline hazelcast-spring/src/test/java/com/hazelcast/spring/cache/TestJCacheCacheManager.java]: shouldn't  the lifecycle be driven by Spring container? I know it's not related to this changeset, but I would expect container to manage lifecycle of Hazelcast instance here.

File: hazelcast-spring/src/test/java/com/hazelcast/spring/cache/TestJCacheCacheManager.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    @Before
+    public void reset() {
+        Iterable<String> cacheNames = springCacheManager.getCacheManager().getCacheNames();
+        for (String cacheName :  cacheNames) {
+            springCacheManager.getCacheManager().getCache(cacheName).removeAll();
```
Comment (jerrinot) [inline hazelcast-spring/src/test/java/com/hazelcast/spring/cache/TestJCacheCacheManager.java]: why is this needed?

Comment (mustafaiman) [inline hazelcast-spring/src/test/java/com/hazelcast/spring/cache/TestJCacheCacheManager.java]: caches are created once before tests start. therefore caches may have leftover entries from other tests. each test should run with empty caches

File: hazelcast-spring/pom.xml (hunk 1 above)
Comment (mustafaiman) [inline hazelcast-spring/pom.xml]: Spring 3 already crashes with JCache 1.0.0. I do not think we have to give users a hint about that.

File: hazelcast-spring/src/test/java/com/hazelcast/spring/cache/JCacheCacheManagerTest.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    @Resource(name = "cacheManager")
+    private JCacheCacheManager springCacheManager;
+
+    @Resource(name = "cacheManager2")
+    private CacheManager cacheManager2;
+
+    @BeforeClass
+    @AfterClass
+    public static void start() {
+        Hazelcast.shutdownAll();
```
Comment (jerrinot) [inline hazelcast-spring/src/test/java/com/hazelcast/spring/cache/JCacheCacheManagerTest.java]: I still do not understand this one. Why is it needed? Hazelcast instance is started by the container, shouldn't the container stop it on its destruction?

Comment (mustafaiman) [inline hazelcast-spring/src/test/java/com/hazelcast/spring/cache/JCacheCacheManagerTest.java]: Sorry, I should have missed it when you first asked.
Sure Spring container stops the instance on its destruction. That statement is to strictly ensure that instances are closed in any failure or success scenario. If `@AfterClass` annotation is removed, behavior will not change. You too can verify that. However I prefer keeping the statement as it is present in all other Spring tests.
I can remove the annotation if you still think I should. What do you think?

Comment (mesutcelik) [inline hazelcast-spring/src/test/java/com/hazelcast/spring/cache/JCacheCacheManagerTest.java]: It seems `@AfterClass` could be removed.

**Comment by mesutcelik** [comment]
:+1: with minor comment.

File: hazelcast-spring/src/test/java/com/hazelcast/spring/cache/JCacheCacheManagerTest.java (hunk 6 above)
Comment (jerrinot) [inline hazelcast-spring/src/test/java/com/hazelcast/spring/cache/JCacheCacheManagerTest.java]: If an instance is not closed -> it's a bug -> we want to know it -> test failure is OK to me. 
Did I miss anything?

**Comment by jerrinot** [comment]
:+1: 

What about a documentation? CC @Serdaro ?

**Comment by mustafaiman** [comment]
Once it is merged, documentation will follow!

**Comment by jerrinot** [comment]
awesome!
